### PR TITLE
Raise Exceptions when js2python is applied to javascript errors

### LINKED ISSUE #888: JsProxy does not handle Error correctly
Example:
```python
from js import Error
e = Error.new("Hi")
raise e
```
Raises: `TypeError: exceptions must derive from BaseException`.
This came up in https://github.com/iodide-project/pyodide/pull/880#pullrequestreview-555341317. I will try to fix this.

### PR DESCRIPTION
**Pull request by hoodmane**
This is intended to resolve #888, necessary for PR #880 to work correctly.
Work in progress.

Currently there is a serious issue: if the Javascript `Error` was created without `new`, this will crash the Pyodide runtime. I have absolutely no idea why.

### CODE REVIEW

**Comment by hoodmane** [comment]
I think this is ready now @oeway.

**Review by oeway — COMMENTED** [review]
Looks good!

I have two questions related to this: 1) have you figured why it's crashing when using Error without new? 2) if we do a round trip from js to python then back to js, will we get the same error object?

File: src/type_conversion/jsproxy.c [hunk 1]
```diff
@@ -544,8 +645,54 @@ JsProxy_AsJs(PyObject* x)
...
+
+  module = PyImport_ImportModule("asyncio");
+  if (module == NULL) {
+    goto fail;
+  }
+
+  asyncio_get_event_loop = PyObject_GetAttrString(module, "get_event_loop");
+  if (asyncio_get_event_loop == NULL) {
+    goto fail;
+  }
```
Comment (rth) [inline src/type_conversion/jsproxy.c:679]: That's unrelated to error conversions, isn't it?

Also some `_Py_IDENTIFIER` in the beginning of this file.

Comment (hoodmane) [reply src/type_conversion/jsproxy.c:679]: @oeway Yes, I figured that out and fixed it. I was incorrectly using a C api and it caused crazy undefined behavior. It should have always crashed, I got unlucky with this `Error` vs `new Error` thing.

Comment (hoodmane) [reply src/type_conversion/jsproxy.c:679]: Right, I'm not that great at git and this is bleedover from the `await_jsproxy` PR #880.

**Comment by hoodmane** [comment]
>1) have you figured why it's crashing when using Error without new? 
Yes I was incorrectly using a C Python API causing undefined behavior. This particular symptom was bad luck. It is fixed.

> 2) if we do a round trip from js to python then back to js, will we get the same error object?
Yes. See test here:
https://github.com/iodide-project/pyodide/blob/ad90a07b5ea6003e2947f0486641292cb53ac42a/src/tests/test_pyodide.py#L41

**Comment by oeway** [comment]
Great, thank you!

**Comment by hoodmane** [comment]
@rth I think this is ready to be merged unless you have further concerns?

File: src/type_conversion/jsproxy.c [hunk 2]
```diff
@@ -544,8 +638,42 @@ JsProxy_AsJs(PyObject* x)
...
+  _Exc_JsException.tp_base = (PyTypeObject*)PyExc_Exception;
+
+  PyObject* module;
+  PyObject* exc;
+
+  module = PyImport_ImportModule("pyodide");
+  if (module == NULL) {
+    goto fail;
+  }
+  if (PyObject_SetAttrString(module, "JsException", Exc_JsException)) {
```
Comment (rth) [inline src/type_conversion/jsproxy.c:668]: Could you please add a comment why we need to do this?

**Review by rth — COMMENTED** [review]
Thanks for working on this @hoodmane and thank you for the review @oeway !

Two comments below, otherwise LGTM.

Please also add an entry to `docs/changelog.md`.

File: src/tests/test_pyodide.py [hunk 3]
```diff
@@ -20,3 +23,39 @@ def test_find_imports():
         )
     )
     assert set(res) == {"numpy", "scipy", "six", "matplotlib"}
+
+
+def test_javascript_error(selenium):
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply src/tests/test_pyodide.py:28]: Moved to `test_type_conversions.py`.

File: src/type_conversion/jsproxy.c (hunk 2 above)
Comment (hoodmane) [reply src/type_conversion/jsproxy.c:668]: Yes. (This is in case someone wants to catch `JsException`.)

**Comment by hoodmane** [comment]
I also added an entry for `JsException` in the api docs. To be clear, I'm envisioning code like:

```pyodide
from js import js_async_func
from pyodide import JsException
async def wrapper(arg):
   try:
      await js_async_func(arg)
   catch JsException:
      raise ValueError(f"Unknown thunk {arg}")
```

**Comment by rth** [comment]
Could you please resolve conflicts following #909 ? Otherwise should be good to merge assuming CI passes.

I have a doubt, would this address https://github.com/iodide-project/pyodide/issues/636 or not?

**Comment by hoodmane** [comment]
This does not fix #636. To do this we should:.
* Add `hiwire_setexc(id)`: this will call `PyExc_SetObject(JsProxy_new_error(id))`.
* Adjust `hiwire_call` and `hiwire_call_member` to use try + catch and call `hiwire_setexc(hiwire_new(err))` for any caught error.

Once this is merged I can implement that. 

Error conversion in the other direction is also not very good: currently if python code throws an error, it prints stuff to stdout directly, but it should instead be possible to catch the error in javascript. To get the errors printed to stdout, we should explicitly catch them and call some separate print function.

**Comment by hoodmane** [comment]
I think the title "Raise Exceptions when js2python is applied to javascript errors" is misleading: `js2python` doesn't raise an error at all, it returns an exception that can be raised. These are very different things:

f raises an error:
```python
def f():
  raise Exception("hi")
```

f returns an error:
```python
def f():
  return Exception("hi")
```

**Comment by rth** [comment]
I had to shorten it as it was too long for a commit message. Didn't mean to rename the issue only the commit. I agree that's not ideal.

**Comment by hoodmane** [comment]
Thanks!

**Pull request merged**